contrib/selinux/syslog_ng.sh

```diff
@@ -125,39 +127,338 @@ setup_vars() {
+		else
+			return 1
+		fi
+	else
+		return 1
+	fi
+}
+
+
+validate_plugin_metadata() {
+	local plugin_dir PLUGIN_NAME PLUGIN_FEATURE PE_SUPPORT_ADDED_VERSION OSE_SUPPORT_ADDED_VERSION
+
+	plugin_dir="${1}"
+	PLUGIN_NAME="${2}"
+	PLUGIN_FEATURE="${3}"
+	PE_SUPPORT_ADDED_VERSION="${4}"
+	OSE_SUPPORT_ADDED_VERSION="${5}"
+
+	if [[ ( -z "${PLUGIN_NAME}" || -z "${PLUGIN_FEATURE}" ) || ( -z "${PE_SUPPORT_ADDED_VERSION}" && -z "${OSE_SUPPORT_ADDED_VERSION}" ) ]]; then
+		echo "Invalid plugin was encountered in directory: '${plugin_dir}'!" >&2
+		exit 243 
+	elif [[ ( "${PLUGIN_FEATURE}" == "PE_ONLY" && -z "${PE_SUPPORT_ADDED_VERSION}" ) || ( "${PLUGIN_FEATURE}" == "OSE_ONLY" && -z "${OSE_SUPPORT_ADDED_VERSION}" ) ]]; then
+		echo "Invalid plugin was encountered in directory: '${plugin_dir}'!" >&2
+		exit 242
+	elif [[ ( "${PLUGIN_FEATURE}" == "ALL" ) && ( -z "${PE_SUPPORT_ADDED_VERSION}" || -z "${OSE_SUPPORT_ADDED_VERSION}" ) ]]; then
+		echo "Invalid plugin was encountered in directory: '${plugin_dir}'!" >&2
+		exit 241
+	fi
+}
+
+
+run_plugin_installers() {
+	local choice
+
+	while read plugin_dir; do
+		PLUGIN_NAME=
+		PLUGIN_FEATURE=
+		PE_SUPPORT_ADDED_VERSION=
+		OSE_SUPPORT_ADDED_VERSION=
+
+		source "${plugin_dir}/include/include.sh"
+		validate_plugin_metadata "${plugin_dir}" "${PLUGIN_NAME}" "${PLUGIN_FEATURE}" "${PE_SUPPORT_ADDED_VERSION}" "${OSE_SUPPORT_ADDED_VERSION}"
+
+		choice=$( echo -ne "${PLUGIN_DB}" | fgrep "${plugin_dir}:" | fetch_plugin_choice )
+		if [ "x${choice}x" == "xYx" ] ; then
+			if plugin_install_precheck "${PLUGIN_FEATURE}" "${PE_SUPPORT_ADDED_VERSION}" "${OSE_SUPPORT_ADDED_VERSION}"; then
+				if already_loaded; then
+					echo "The SELinux module for '${PLUGIN_NAME}' is already loaded! Skipping install..." >&2
+				else
+					"${plugin_dir}/setup.sh" --install-dir "${INSTALL_PATH}" --os "${OS_VERSION}"
+				fi
+			fi
+		fi
+	done < <( dump_PLUGIN_DB | fetch_plugin_dir )
+}
+
+
+run_plugin_callbacks() {
+	local choice buffer
+	while read plugin_dir; do
+		PLUGIN_NAME=
+		PLUGIN_FEATURE=
+		PE_SUPPORT_ADDED_VERSION=
+		OSE_SUPPORT_ADDED_VERSION=
+
+		source "${plugin_dir}/include/include.sh"
+		validate_plugin_metadata "${plugin_dir}" "${PLUGIN_NAME}" "${PLUGIN_FEATURE}" "${PE_SUPPORT_ADDED_VERSION}" "${OSE_SUPPORT_ADDED_VERSION}"
+
+		choice=$( dump_PLUGIN_DB | fgrep "${plugin_dir}:" | fetch_plugin_choice )
+		if [ "x${choice}x" == "xYx" ] ; then
+			for type in te if fc; do
+				buffer=$( return_files "${type}" )
+				[ -n "${buffer}" ] && cat >> "syslog_ng.${type}" <<<"${buffer}"
+			done
+		fi
+	done < <( dump_PLUGIN_DB | fetch_plugin_dir )
+}
+
+
+remove_plugin_modules() {
+	while read plugin_dir; do
+		PLUGIN_NAME=
+		PLUGIN_FEATURE=
+		PE_SUPPORT_ADDED_VERSION=
+		OSE_SUPPORT_ADDED_VERSION=
+
+		source "${plugin_dir}/include/include.sh"
+		validate_plugin_metadata "${plugin_dir}" "${PLUGIN_NAME}" "${PLUGIN_FEATURE}" "${PE_SUPPORT_ADDED_VERSION}" "${OSE_SUPPORT_ADDED_VERSION}"
+
+		if already_loaded; then
```

Comment:
Sorry if I miss something, but I don't find any definition of `already_loaded`, I've checked `common.sh.inc` too.

Reply:
Yes, this function comes from the plugin's include file. As the main installer script has no way of knowing how a plugin's SELinux module will be called.
So it sources code (2 functions) and metadata from the plugin's subdirectory to 
* get a list of files that need to be included in the syslog policy (ie. syslog-ng needs to communicate with a separate component)
* check if the plugin's separate SELinux module has already been loaded or not

And the metadata is sourced so that the name and feature or version constraints can be checked.

Reply:
Thanks for the explanation!

A hard-coded function name which is defined by a plugin IMHO is not a good practice.
It's hard to understand and hard to read. As a plugin developer I don't know what to interfaces to implement, what should be the implementation doing.
I've just discovered that these functions don't have any implementation and they will be defined the a plugin.

On high-level I see the following information about plugins:
- plugins have a  have a directory format (see `get_plugin_list`): a plugin is in a directory in
- plugins have an install script `setup.sh` with defined parameters that places
- plugins can have an additional install step, which is done by the "plugin_callbacks" which is the `return_files` function defined by the plugin.
- plugins must define a `already_loaded` function which checks that the plugin files are installed.

Just an observation:
I think you can replace `return_files` definition with a hard-coded script name, that the plugins must implement (i.e. an interface), e.g. `${plugin_dir}/callback.sh`

And you can replace the `already_loaded` with a `${plugin_dir}/is_installed.sh` or the `setup.sh` script itself can do the check before the actual information and if it's already installed it will return with an error code.

Reply:
Seems like a good idea. I will try to implement it...